### PR TITLE
Add option to return matched/unmatched/unexpected in 'load_state_dict' fixed #9530

### LINKED ISSUE #9530: [feature request] Add option to return matched / unmatched / unexpected in `load_state_dict`
Currently, loading checkpoints using `strict=False` can be fairly error-prone. An extra level (for example with a `nn.DataParallel`) could lead to completely missing loading all the weights, and there is no sign that this has happened.

### Proposal: 
A way to enable user-code to verify for possible mistakes would be to add an option to `load_state_dict` to return matched / unmatched / unexpected keys.
This would allow the user to easily inspect if the loading is actually doing what is expected.

Returning those values is only needed if `strict=False`, because in `strict=True` mode an error is raised if the state dict doesn't match exactly.

A possible API:
```python
def load_state_dict(self, state_dict, strict=True, return_unmatched=False):
    ...
    if return_unmatched:
        return matched, unmatched, unexpected
```

Thoughts?

### PR DESCRIPTION
**Pull request by bhushan23**
Add option to return matched/unmatched/unexpected in 'load_state_dict'

Adding test case in TestNN.test_load_state_dict 
Implements #9530

### CODE REVIEW

**Comment by bhushan23** [comment]
builds are failing with following error
In file included from /var/lib/jenkins/workspace/torch/../aten/src/ATen/cuda/CUDAContext.h:4:0,
02:05:03                  from /var/lib/jenkins/workspace/torch/csrc/cuda/comm.h:5,
02:05:03                  from /var/lib/jenkins/workspace/torch/csrc/autograd/functions/comm.cpp:6:
02:05:03 /var/lib/jenkins/workspace/torch/../aten/src/ATen/cuda/CUDAStream.h:6:30: fatal error: cuda_runtime_api.h: No such file or directory
02:05:03  #include "cuda_runtime_api.h"

**Comment by bhushan23** [comment]
Added test is asserting in one of the build. Need to update the tests, will do that today but this does not blocks code review of load_state

**Comment by apaszke** [comment]
Uh I missed the comments in the original issue. Let's finalize the discussion on the interface before we merge this.

**Review by apaszke — CHANGES_REQUESTED** [review]
Didn't take a look at the code yet. We should decide on the API first.

**Comment by bhushan23** [comment]
@fmassa please review the prototype 
Please don't do in depth code review. This is just to verify and confirm approach

**Comment by bhushan23** [comment]
@fmassa @soumith can you review the prototype?

**Comment by yf225** [comment]
@fmassa @soumith Pinging for prototype review

File: torch/nn/modules/module.py
Comment (fmassa) [inline torch/nn/modules/module.py]: I think we have decided that it's better not to return the keys conditionally on an argument?

File: torch/nn/modules/module.py
Comment (fmassa) [inline torch/nn/modules/module.py]: Can't we always compute this?

**Review by fmassa — CHANGES_REQUESTED** [review]
I think this looks ok, but from what I understood we would be avoiding returning the `matched` / `unmatched` / `unexpected` ?

**Comment by bhushan23** [comment]
@fmassa you are right, removing returning keys from load_state_dict

Regarding, computing keys always, we don't need to.
We will compute keys only if strict is passed

**Comment by bhushan23** [comment]
this is weird, checking for difference before loading leads to some unmatched keys but doing within load and after does not. @fmassa do you know why? 
Have updated change with this.

**Comment by bhushan23** [comment]
@apaszke are you also reviewing?

File: torch/nn/modules/module.py
Comment (fmassa) [inline torch/nn/modules/module.py]: looks like this block is duplicated?

File: torch/jit/__init__.py
Comment (fmassa) [inline torch/jit/__init__.py]: @zou3519 does the JIT know how to compile functions using python `set`?

**Review by fmassa — CHANGES_REQUESTED** [review]
There is a duplicated error message that should be fixed.
Also, I'm not sure that the JIT supports that function, but I'll let the others comment on that.

Apart from that, it generally looks good to me

File: torch/jit/__init__.py
Comment (zou3519) [reply torch/jit/__init__.py]: What do you mean by "compile functions using python set"?

Comment (fmassa) [reply torch/jit/__init__.py]: I probably don't know what I'm saying, but what's the purpose of `_compiled_methods_whitelist` in the JIT?

File: torch/nn/modules/module.py
Comment (apaszke) [inline torch/nn/modules/module.py]: I'm not 100% sure if we really want to have this function. The beginning could be simplified to
```py
local_keys = set(self.state_dict())
state_keys = set(state_dict)
```
and then it becomes a few lines of very straightforward Python code. To be honest it would be harder for me to find this function to the docs, and make sure that I assign the three lists we return in the correct order, than to write them out myself...

**Review by apaszke — CHANGES_REQUESTED** [review]
I have serious doubts about this functionality. It doesn't seem worth having in the core for me. Everyone can implement them in their projects in case they need it, and it might even end up being faster in case you don't need one of the three sets we compute.

**Comment by bhushan23** [comment]
@apaszke 
>> Everyone can implement them in their projects in case they need it, 
This this API, one can quickly query state of the state_dict (and one get full picture)
And doing this in code is a bit messy, why should we force everyone to find the difference keys manually every time they need it?
>> and it might even end up being faster in case you don't need one of the three sets we compute.
I agree here. 
How about passing keys of interest and we only compute those?
But, I guess, if someone is using diff_state_dict, they are okay with extra time.

Also, now returning keys as set instead of converting into list. 
Keys are supposed to be set, then why waste time in un-necessary conversion into list.

**Comment by apaszke** [comment]
I'm just not convinced that it's a good API. It's a very simple thing, and we're making the interface of a module unnecessarily complex. Also, as I mentioned, returning 3 values of the same type makes it extremely easy to mix them up, and there's really no way to remember what this function does without looking it up in the docs every time. On the other hand, if you had this code in your project, it would be much harder to make a mistake.

**Comment by bhushan23** [comment]
@apaszke I get your point. Returning 3 different keys can lead to confusion. 
How about returning dictionary? Key will be string (matched / unmatched / unexpected) and value will be set of those keys.

**Comment by fmassa** [comment]
After a lot of discussion, we think that it's better to let the user handle this manually, and use `strict=True` mode when loading.

Thanks a lot for this PR!

**Pull request closed without merging**
